Create inline_link_in_paragraph_is_distinguishable_ be4d0c.md (#1010)
act-rules/act-rules.github.io: _rules/inline-link-in-paragraph-has-distinguishable-style-not-based-on-color-alone-88407d.md

---
id: 88407d

> name: Inline link has distinguishable style or content not based on color alone

Reviewer: How can a distinction based on content rely on color? The distinction based on color has to do with styles more than with content. So, I propose flipping around the title a little bit so that it reads clearer

```suggestion
name: Inline link has distinguishable style not based on color alone
or content that indicates it is a link
```

rule_type: atomic
description: |
  This rule checks that inline links are distinguishable from the surrounding text through a difference in style not based on color alone or have content that indicates they are links.

> accessibility_requirements:

Reviewer 2: Do atomic rules not include any accessibility requirements?

> accessibility_requirements:

Author: This (partially) covers SC 1.4.1. But it would not support the fail-on-fail relationship that rules need to have. Therefore the accessibility requirements end up only in the composite rule.

input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgments:
  authors:
    - Brian Bors
    - Carlos Duarte
    - Wilco Fiers
---

## Applicability

This rule applies to any [visible][] HTML element that is a [semantic link][], for which all the following is true:

- **link text**: the element has [visible][] [text nodes][text node] as [descendants][descendant] in the [flat tree][]; and
- **non-link line text**: the element is [rendered on a line][] containing [visible][] [text nodes][text node] that are not [descendants][descendant] in the [flat tree][] of a [semantic link][]; and

> - **different color**: the element's [foreground color][] and the [foreground color][] of the **non-link line text** elements have a [highest possible contrast][] that is at least 3.0:1, or the element's [background color][] and the [background color][] of the **non-link line text** elements have a [highest possible contrast][] that is at least 3.0:1.

Author: This rule is (partially) checking SC 1.4.1: Use of Color. The rule looks at inline links that are distinguishable by color in order to check if they have another distinguishing style (other than color).
If the link is not distinguishable by color, then this SC does not apply.

@WilcoFiers and @EmmaJP do you still think that color should be dropped from the applicability?

> - **different color**: the element's [foreground color][] and the [foreground color][] of the **non-link line text** elements have a [highest possible contrast][] that is at least 3.0:1, or the element's [background color][] and the [background color][] of the **non-link line text** elements have a [highest possible contrast][] that is at least 3.0:1.

Author: After reading the following comment I understand where this is coming from, and I guess the rule is going to go through a big simplification.

> - **different color**: the element's [foreground color][] and the [foreground color][] of the **non-link line text** elements have a [highest possible contrast][] that is at least 3.0:1, or the element's [background color][] and the [background color][] of the **non-link line text** elements have a [highest possible contrast][] that is at least 3.0:1.

Reviewer 2: Perhaps there should also be a passing example where the luminescence contrast is the additional distinguishing style?

## Expectation

For each test target, at least one of the following is true when it is neither [focused][] nor [hovered][]:

- **distinguishing style**: the element has a [distinguishable style][] not based on color alone from the **non-link line text** elements; or

> - **distinguishing content**: the element has content (such as an image or text) located in the proximity of the element (i.e. there is no other content visually in between this content and the target element), that indicates the test target is a link.

Reviewer: ```suggestion
- **distinguishing content**: the element has adjacent content (such as an image or text) that indicates the test target is a link.
```

	}
</style>
<p>Read about WAI on the <a href="http://w3.org/WAI">WAI webpage</a>.</p>
```

#### Passed Example 7

This link, that is distinguishable by color from the other text in the same line, has an icon that makes it distinguishable as a link.

```html
<style>
	p {
		color: black;
		background-color: white;
	}
	a.test {
		color: blue;
		background-color: white;
		text-decoration: none;
	}
</style>
<p>
	Read about WAI on the
	<a class="test" href="http://w3.org/WAI">WAI webpage <img src="/test-assets/be4d0c/icon.png" alt="" /></a>.
</p>
```

#### Passed Example 8

This link, that is distinguishable by color from the other text in the same line, has text that makes it distinguishable as a link.

```html
<style>
	p {
		color: black;
		background-color: white;
	}
	a.test {
		color: blue;
		background-color: white;
		text-decoration: none;
	}
</style>
<p>Read about WAI on the <a class="test" href="http://w3.org/WAI">WAI webpage by following this link</a>.</p>
```

### Failed

#### Failed Example 1

This link, that is distinguishable by color from the other text in the same line, has no other visual cues of being recognized as a link with the underline removed.

> 

Author: Will update the example(s)

> 

Reviewer 2: Perhaps also need to be clearer in the description that the colour difference between blue and black is insufficient on its own because it is less than a 3:1 contrast.

---
id: 88407d

> name: Inline link has distinguishable style not based on color alone or content that indicates it is a link

Reviewer 3: I feel it,s not fully clear that the "or" applies to "distinguishable style" (rather than to "based on color").
```suggestion
name: Inline link has either distinguishable style not based on color alone or content that indicates it is a link
```

- **link text**: the element has [visible][] [text nodes][text node] as [descendants][descendant] in the [flat tree][]; and
- **non-link line text**: the element is [rendered on a line][] containing [visible][] [text nodes][text node] that are not [descendants][descendant] in the [flat tree][] of a [semantic link][]; and

> - **different hue**: the element's [foreground color][] and the [foreground color][] of the **non-link line text** elements have a [different hue][], or the element's [background color][] and the [background color][] of the **non-link line text** elements have a [different hue][].

Reviewer 3: Since we are actually comparing sets of colors, shouldn't it be plurals ("foreground colors", …)?

## Expectation

> For at least one element of the set comprised by the test target and all of its [visible][] [descendant][] elements in the [flat tree][], at least one of the following is true when the element is neither [focused][] nor [hovered][]:

Reviewer 3: I'm not sure that "all of its visible descendants" is good enough:
```html
<p>Lorem <a href="#" class="no-style">A very long link name with <span class="border">a</span> single letter with border</a> Ipsum</p>
```
That would pass the rule. Does that pass the SC? I don't think so…
OTOH
```html
<p>Lorem <a href="#" class="no-style">A <span class="border">very long link name with a single letter without border </span></a> Ipsum</p>
```
could arguably satisfy the SC just because the indication is taking so much of the actual click target that it's OK in virtually all practical cases 🤔

Reviewer 4: Similar to how the text contrast rule accepts a single foreground pixel with sufficient contrast even if _all_ the other foreground pixels have insufficient contrast, I believe this rule is really doing the best it can. Most importantly, it preserves a non-controversial fail-on-fail relation to the success criterion.

For each test target, at least one of the following is true:

> - **distinguishing style**: the element has a [distinguishable style][] not based on color alone from the **non-link line text** elements; or

Reviewer 3: What if the distinguishing style is applied to an ancestor?
`<p>Lorem <span class="border"><span><span><a href="#" class="no-style">link</a></span></span></span> Ipsum</p>`
is currently failing the rule but passing the SC.
And the latest assumption is accepting styling on ancestors, so doesn't catch that.

Author: The only way I see to tackle this is to extend the assumption to include ancestors...

- **distinguishing content**: the element has content (such as an image or text), located in the proximity of the element (i.e. there is no other content visually in between this content and the target element), that indicates the test target is a link.

## Assumptions

> - The 3:1 contrast difference between text color is the minimum required to meet WCAG 2.0. This value is part of [technique G183](https://www.w3.org/WAI/WCAG21/Techniques/general/G183) but is not specified in the [1.4.1 success criterion](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html).

Reviewer 3: I still feel that doesn't really read as an assumption.
As far as I can tell, the assumption is that "something with a contrast lower than 3:1 is not concerned by 1.4.1".

This does tie a bit with https://github.com/act-rules/act-rules.github.io/pull/1010#discussion_r459320027 If the link does not have that contrast ratio, which SC is it failing (if any)? I feel it should definitely fail something, but I can't find what…

Author: Rephrased it. I think it's easier to understand now.

- Any change in font is sufficiently distinguishable, and fonts are loaded when they are present.
- The same foreground color and the same background color is used by all the visible text nodes of the _ancestor_ element that are not part of the target element, otherwise color can not be a distinguishing factor.
- The same `box-shadow` is used by all the visible text nodes of the _ancestor_ element that are not part of the target element, otherwise `box-shadow` can not be a distinguishing factor.
- The same `border` is used by all the visible text nodes of the _ancestor_ element that are not part of the target element, otherwise `border` can not be a distinguishing factor.

> - The perceived visual styling of elements is the result of CSS styling applied to those elements and their _ancestor_ elements. If, through manipulation of the CSS styles of elements that are not rendered on the same line, an element ends up with a distinguishable style then this rule might fail while the success criteria might still be satisfied.

Reviewer 3: ```suggestion
- The perceived visual styling of elements is the result of CSS styling applied to those elements and their _ancestor_ elements. Through manipulation of the positioning and styles of other elements, an element may end up with a perceived distinguishable style not resulting from its own or ancestors styling. In that case, this rule might fail while the [Success Criterion 1.4.1 Use of Color](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html) might still be satisfied.
```
